grpc_collector_client.go

```diff
@@ -194,15 +194,25 @@ func (client *grpcCollectorClient) makeReportRequest(buffer *reportBuffer) *cpb.
 
 }
 
-func (client *grpcCollectorClient) Report(ctx context.Context, buffer *reportBuffer) (collectorResponse, error) {
-	resp, err := client.grpcClient.Report(ctx, client.makeReportRequest(buffer))
+func (client *grpcCollectorClient) Report(ctx context.Context, req reportRequest) (collectorResponse, error) {
+	if req.grpcRequest == nil {
```

Comment:
Oh I miss generics so much :(

Instead of having a struct with many different types of pointers, what about replacing `reportRequest` with `interface{}` and then casting? I feel like that might be a little bit cleaner.

It's just a thought, I don't feel strongly about it.

Reply:
Yeah, I very nearly said the same thing, but don't feel strongly about it.

Reply:
huh @jmacd I could have sworn I ended up with the struct instead of an interface after a chat with you. Happy to do whatever we think makes more sense.

Reply:
@jmacd I could have sworn that I ended up chosing this approach over the interface after a conversation with you, but if that is not correct and neither of us can rembmer / come up with a good reason for why this appraoch was taken, then I am all for switching to interfaces. Just want to make sure there wasn't a good reason that I'm forgetting.

Reply:
I think I'm going to leave this as is, since the actual requests are all specific the the networks and are not LS code, so making an interface would basically just be a wrapper. ie 
`type reportRequest interface {
      TypedRequest{} interface{}
}`
which i'm not certain how much cleaner that really makes things. Thinking I'll leave as is for now, especially since it sounds like much of this going to change soon anyway.

Reply:
Totally ok with the decision to leave it as is, but just wanted to be clear, because I think we might be talking about a slightly different proposal.

I was suggesting that the function read like this:

```
func (client *grpcCollectorClient) Report(
   ctx context.Context,
   req interface{},
) (collectorResponse, error) {
  protoReq, ok := req.(*cp.ReportRequest)
  if !ok {
    return nil, fmt.Errorf("client / encoder combination not supported.")
  }

  ...
}
```